### PR TITLE
Zuul rate limit

### PR DESCRIPTION
**Pull request by viniciusccarvalho**
Adds ratelimit filter to zuul. 

The filter is enabled when `zuul.ratelimit.enabled` is present.

It has an inmemory and a redis based implementation. The redis implementation will work with multiple instances.

There's two policies supported:
```
zuul:
  ratelimit:
    enabled: true
    policies:
        anonymous:
            refreshInterval: 60
            limit: 1000
        authenticated:
            refreshInterval: 60
            limit: 2000
```
The policies control how we create the buckets. Anonymous will use a global bucket counter, the authenticated uses the security principal as the key, so each authenticated user has its own counter.

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/InMemoryRateLimiter.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *  limitations under the License.
+ *
+ */
+
+package org.springframework.cloud.netflix.zuul.filters.pre.ratelimit;
+
+import java.util.WeakHashMap;
+
+/**
+ * @author Vinicius Carvalho
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/InMemoryRateLimiter.java]: Author at end of javadoc :-)

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/Policy.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	public Policy() {
+	}
+
+	public Policy(Long refreshInterval, Long limit) {
+		this.refreshInterval = refreshInterval;
+		this.limit = limit;
+	}
+
+	public static interface PolicyType {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/Policy.java]: Maybe an enum?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/Rate.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * @author Vinicius Carvalho
+ *
+ * Represents a view of rate limit in a giving time for a user.
+ *
+ * limit - How many requests can be executed by the user. Maps to X-RateLimit-Limit header
+ * remaining - How many requests are still left on the current window. Maps to X-RateLimit-Remaining header
+ * reset - Epoch when the rate is replenished by limit. Maps to X-RateLimit-Reset header
+ */
+@Data
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/Rate.java]: We're phasing lombok out, may as well not include it.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		@Bean
+		public RateLimiter inMemoryRateLimiter() {
+			InMemoryRateLimiter rateLimiter = new InMemoryRateLimiter();
+			return rateLimiter;
+		}
+
+	}
+
+
+	@ConditionalOnBean(name = {"redisTemplate"})
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitConfiguration.java]: Should have conditional on class `RedisTemplate` too, to guard if it's not on the classpath.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import com.netflix.zuul.exception.ZuulException;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Vinicius Carvalho
+ */
+@Component
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java]: Don't rely on classpath scanning. This should be created in an auto configuration class.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+@ConditionalOnProperty("zuul.ratelimit.enabled")
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java]: Move this to the auto configuration class.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java [hunk 7]
```diff
@@ -156,6 +156,9 @@ public boolean shouldFilter() {
 	public Object run() {
 		RequestContext context = RequestContext.getCurrentContext();
 		HttpServletRequest request = context.getRequest();
+		if(context.get("warning") != null){
+			System.out.printf("should not stop here");
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: use logging

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/RateLimitZuulProxyApplicationTests.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		assertNotNull(response.getHeaders().get(RateLimitFilter.Headers.REMAINING));
+		assertNotNull(response.getHeaders().get(RateLimitFilter.Headers.RESET));
+	}
+}
+
+@Configuration
+@EnableAutoConfiguration
+@RestController
+@EnableZuulProxy
+@ComponentScan(basePackageClasses = RateLimitConfiguration.class)
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/RateLimitZuulProxyApplicationTests.java]: Again, don't component scan, auto-config.

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/SampleZuulProxyApplicationTests.java [hunk 9]
```diff
@@ -274,7 +274,7 @@ public MyCommand(int errorCode, String commandKey, RestClient restClient,
 				HttpRequest.Verb verb, String uri, Boolean retryable,
 				MultiValueMap<String, String> headers,
 				MultiValueMap<String, String> params, InputStream requestEntity)
-						throws URISyntaxException {
+				throws URISyntaxException {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/SampleZuulProxyApplicationTests.java]: revert whitespace changes

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/ServletPathZuulProxyApplicationTests.java [hunk 10]
```diff
@@ -100,6 +100,8 @@ public void stripPrefixFalseAppendsPath() {
 @EnableZuulProxy
 class ServletPathZuulProxyApplication {
 
+
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/ServletPathZuulProxyApplicationTests.java]: revert whitespace changes

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+		Policy policy = findRequestPolicy(request);
+		String key = findKey(request);
+		Rate rate = limiter.consume(policy, key);
+		response.setHeader(Headers.LIMIT, rate.getLimit().toString());
+		response.setHeader(Headers.REMAINING, Math.max(rate.getRemaining(), 0) + "");
+		response.setHeader(Headers.RESET, rate.getReset().toString());
+		if (rate.getRemaining() <= 0) {
+			ctx.setResponseStatusCode(429);
+			ctx.put("warning","on");
+			throw new RuntimeException("");
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java]: Either remove `""` or put some information in it.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+		HttpServletRequest request = ctx.getRequest();
+		Policy policy = findRequestPolicy(request);
+		String key = findKey(request);
+		Rate rate = limiter.consume(policy, key);
+		response.setHeader(Headers.LIMIT, rate.getLimit().toString());
+		response.setHeader(Headers.REMAINING, Math.max(rate.getRemaining(), 0) + "");
+		response.setHeader(Headers.RESET, rate.getReset().toString());
+		if (rate.getRemaining() <= 0) {
+			ctx.setResponseStatusCode(429);
+			ctx.put("warning","on");
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java]: probably remove this debugging?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitFilter.java]: Actually, maybe you should put something like `rateLimitExceeded=true` and the routing filters shouldn't run if set.

File: spring-cloud-netflix-core/src/main/resources/META-INF/spring.factories [hunk 13]
```diff
@@ -7,10 +7,13 @@ org.springframework.cloud.netflix.feign.encoding.FeignContentGzipEncodingAutoCon
 org.springframework.cloud.netflix.hystrix.HystrixAutoConfiguration,\
 org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration,\
 org.springframework.cloud.netflix.rx.RxJavaAutoConfiguration,\
-org.springframework.cloud.netflix.metrics.servo.ServoMetricsAutoConfiguration
-
+org.springframework.cloud.netflix.metrics.servo.ServoMetricsAutoConfiguration,\
+org.springframework.cloud.netflix.zuul.filters.pre.ratelimit.RateLimitConfiguration
+org.springframework.cloud.bootstrap.BootstrapConfiguration=\
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/resources/META-INF/spring.factories]: This seems weird, where did that come from?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.data.redis.core.RedisTemplate;
+
+/**
+ * @author Vinicius Carvalho
+ */
+@Configuration
+@EnableConfigurationProperties(RateLimitProperties.class)
+@ConditionalOnProperty("zuul.ratelimit.enabled")
+
+public class RateLimitConfiguration {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/ratelimit/RateLimitConfiguration.java]: Should be `RateLimitAutoConfiguration`.

File: spring-cloud-netflix-core/src/main/resources/META-INF/spring.factories (hunk 13 above)
Comment (viniciusccarvalho) [inline spring-cloud-netflix-core/src/main/resources/META-INF/spring.factories]: From the main branch I guess: https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-core/src/main/resources/META-INF/spring.factories#L10

**Comment by spencergibb** [comment]
Master changes shouldn't show up in your PR. You might need to rebase and force push.

**Comment by spencergibb** [comment]
Closing in favor of #895

**Pull request closed without merging**